### PR TITLE
CountVectorizer

### PR DESCRIPTION
**Pull request by angryjinyan**
CountVectorizerFeaturizer finished

### CODE REVIEW

File: src/Featurizers/CountVectorizerFeaturizer.h [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+    // ----------------------------------------------------------------------
+
+    // MSVC has problems when the definition and declaration are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+
+        using AppearanceMapType = typename std::map<IterRangeType, std::uint32_t, Components::IterRangeComp>;
+
+        AppearanceMapType appearanceMap;
+
+        Strings::Parse<std::string::const_iterator>(
```
Comment (wangyems) [inline src/Featurizers/CountVectorizerFeaturizer.h]: The parse method in execute() needs to be the same as fit() in DocumentStatisticEstimator, let's sync on this afternoon. The execute() in TfidfTransformer is a quick reference for now

File: src/Featurizers/CountVectorizerFeaturizer.h [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+class CountVectorizerTransformer : public StandardTransformer<std::string, SparseVectorEncoding<std::uint32_t>> {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using BaseType                           = StandardTransformer<std::string, SparseVectorEncoding<std::uint32_t>>;
+    using IndexMapType                       = std::unordered_map<std::string, std::uint32_t>;
+    using AnalyzerMethod                     = Components::Details::DocumentStatisticsTrainingOnlyPolicy::AnalyzerMethod;
```
Comment (davidbrownellWork) [inline src/Featurizers/CountVectorizerFeaturizer.h]: >Components::Details::DocumentStatisticsTrainingOnlyPolicy::AnalyzerMethod; [](start = 47, length = 74)

We should not reference members in a details namespace - this enum should be moved up a level.

File: src/Featurizers/CountVectorizerFeaturizer.h [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+        std::move([&map](void) ->  IndexMapType & {
+            if (map.size() == 0) {
+                throw std::invalid_argument("Index map is empty!");
+            }
+            return map;
+        }()
+        )),
+    _binary(std::move(binary)),
+    _lower(std::move(lower)),
+    _analyzer(std::move(analyzer)),
```
Comment (davidbrownellWork) [inline src/Featurizers/CountVectorizerFeaturizer.h]: >_analyzer(std::move(analyzer)), [](start = 4, length = 31)

Validate.

File: src/Featurizers/CountVectorizerFeaturizer.h [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+                    Microsoft::Featurizer::Strings::ParseNgramCharwb<std::string::const_iterator, std::function<bool (char)>>(
+                        input,
+                        [] (char c) {return std::isspace(c);},
+                        _ngram_min,
+                        _ngram_max,
+                        callback
+                    );
+                };
+            }
+            return parseFunc;
```
Comment (davidbrownellWork) [inline src/Featurizers/CountVectorizerFeaturizer.h]: I didn't expect to see this code here - isn't this handled by DocumentStatisticsEstimator?

File: src/Featurizers/CountVectorizerFeaturizer.h [hunk 5]
```diff
@@ -0,0 +1,481 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
```
Comment (yuslepukhin) [inline src/Featurizers/CountVectorizerFeaturizer.h]: >ragma once [](start = 2, length = 10)

Add this file to a CMakeFile

Comment (angryjinyan) [reply src/Featurizers/CountVectorizerFeaturizer.h]: added

**Pull request closed without merging**